src/supervised/supervised.jl

```diff
@@ -36,141 +36,150 @@ islipschitzcont_deriv(::SupervisedLoss) = false
 
 # --------------------------------------------------------------
 
-@inline function value(loss::SupervisedLoss, target::AbstractVecOrMat, output::AbstractVecOrMat)
+@inline function value(loss::SupervisedLoss, target::AbstractArray, output::AbstractArray)
     buffer = similar(output)
     value!(buffer, loss, target, output)
 end
 
-@inline function deriv(loss::SupervisedLoss, target::AbstractVector, output::AbstractVecOrMat)
+@inline function deriv(loss::SupervisedLoss, target::AbstractArray, output::AbstractArray)
     buffer = similar(output)
     deriv!(buffer, loss, target, output)
 end
 
-@inline function grad(loss::SupervisedLoss, target::AbstractMatrix, output::AbstractVecOrMat)
-    buffer = similar(output)
-    grad!(buffer, loss, target, output)
-end
+# @inline function grad(loss::SupervisedLoss, target::AbstractMatrix, output::AbstractVecOrMat)
+#     buffer = similar(output)
+#     grad!(buffer, loss, target, output)
+# end
 
 # --------------------------------------------------------------
-
-function value!(buffer::AbstractVector, loss::SupervisedLoss, target::AbstractVector, output::AbstractVector)
-    n = length(output)
-    @_dimcheck length(target) == n && size(buffer) == size(output)
-    @simd for i = 1:n
-        @inbounds buffer[i] = value(loss, target[i], output[i])
-    end
-    buffer
-end
-
-function deriv!(buffer::AbstractVector, loss::SupervisedLoss, target::AbstractVector, output::AbstractVector)
-    n = length(output)
-    @_dimcheck length(target) == n && size(buffer) == size(output)
-    @simd for i = 1:n
-        @inbounds buffer[i] = deriv(loss, target[i], output[i])
+# value!, deriv!
+# `output` can have more dimensions than `target`, in which case do broadcasting
+
+@generated function value!{T,N,Q,M}(
+        buffer::AbstractArray,
+        loss::SupervisedLoss,
+        target::AbstractArray{Q,M},
+        output::AbstractArray{T,N}
+    )
+    quote
+      @_dimcheck size(buffer) == size(output)
+      @nexprs $N (n)->(s_n = size(output,n))
+      @nexprs $M (n)->@_dimcheck(size(target,n) == s_n)
+      @simd for I in CartesianRange(@ntuple($N,n->(1:s_n)))
+          @nexprs $N n->(i_n = I[n])
+          @inbounds @nref($N,buffer,i) = value(loss, @nref($M,target,i), @nref($N,output,i))
```

Comment:
I am trying to parse this in my head. What if `M` is larger than `N`? there are only `N` `i`'s.
yet `@nref($M,target,i)` will exptect `M` `i`'s. I think?